File: server/game/modules/tic-tac-toe/tic-tac-toe-service/Middleware/ServiceWhitelistMiddleware.cs
using Microsoft.Extensions.Primitives;
using Serilog;
using Serilog.Context;

public class ServiceWhitelistMiddleware
{
    private readonly RequestDelegate _next;
    private readonly string[] _whitelistedServices;
    private readonly string _serviceHeaderKey;

    public ServiceWhitelistMiddleware(RequestDelegate next, IConfiguration configuration)
    {
        _next = next;

        _serviceHeaderKey = Environment.GetEnvironmentVariable("INTERNAL_SERVICE_HEADER")
                            ?? configuration["ServiceSettings:InternalServiceHeader"]
                            ?? "internal-service";

        _whitelistedServices = configuration["ServiceSettings:WhitelistedServices"]?.Split(',') ?? ["game_gateway_service"];
    }

    public async Task InvokeAsync(HttpContext context)
    {
        using (LogContext.PushProperty("Context", "ServiceWhitelist.Middleware"))
        {
            if (!context.Request.Headers.TryGetValue(_serviceHeaderKey, out StringValues serviceHeader))
            {
                Log.Warning("Access forbidden: No service header provided.");
                context.Response.StatusCode = 403;
                await context.Response.WriteAsync("Access forbidden: No service header provided.");
                return;
            }

            var serviceHeaderValue = serviceHeader.ToString();

            Log.Information("Service request from {ServiceHeader}", serviceHeaderValue);

            if (_whitelistedServices.Contains(serviceHeaderValue))
            {
                Log.Information("Access granted for service: {ServiceHeader}", serviceHeaderValue);
                await _next(context);
            }
            else
            {
                Log.Warning("Access forbidden: Service {ServiceHeader} is not allowed.", serviceHeaderValue);
                context.Response.StatusCode = 403;
                await context.Response.WriteAsync("Access forbidden: Service not allowed.");
            }
        }
    }
}

File: server/auth/auth-repository-service/src/middleware/serviceWhitelistMiddleware.js
const WHITELISTED_SERVICES = [
  "auth_service",
  "game_gateway_service",
  "communication_gateway_service",
];

export const serviceWhitelistMiddleware = (req, res, next) => {
  const serviceHeaderKey = process.env.INTERNAL_SERVICE_HEADER;
  const serviceHeader = req.headers[serviceHeaderKey];

  console.log("Request from:", serviceHeader || req.ip);

  if (WHITELISTED_SERVICES.includes(serviceHeader)) {
    next();
  } else {
    res.status(403).json({ message: "Access forbidden: Service not allowed" });
  }
};
